### ISSUE TITLE
Implement [CARRIERAUTO], i.e. auto-approve and -remove carrier objects

### ISSUE THREAD

**Issue by arnoldnipper** [peeringdb/peeringdb#1335]
**Is your feature request related to a problem? Please describe.**
The `carrier` object was implemented lately. However, there is no clear policy **how and when to approve** a carrier.

**Who is affected by the problem?**

- Everyone requesting a `carrier` object
- @peeringdb/ac 

**What is the impact?**

- automation
- faster turn around of requests

**Are there security concerns?**
no

**Are there privacy concerns?**
no

**Describe the solution you'd like**

- add a `fac_count` field to the `carrier` object. Default is 0
- every carrier presence at a facility has to be approved by a facility admin. This is enough source of trust
- Show the `carrier` object iff `fac_count` > 0

**Do you think this feature will require a formal design?**
IMO, above algorithm should do

**What is the proposed priority?**
In a constant attempt to automate, put it on top of the priority list

**Comment by mustangthz** [comment]
I don't think it should auto approve.  I think it should send an email to the FAC admins to approve.

**Comment by arnoldnipper** [comment]
> I don't think it should auto approve. I think it should send an email to the FAC admins to approve.

I should have been more verbose. There are two steps.

1. create the `carrier` object
2. make the `carrier`object **visible/invisible**

My proposal is to fully automate the `carrier` object handling

1. If a `carrier` request comes in, auto-create the object, but keep it invisible as its `fac_count` is 0. Iow: `status` is `pending`
2. If a carrier operator adds a presence to a facility
2.1 create a `carrierfac` object but keep it invisible, Iow: `status`is pending
2.2 send a notification to all facility admins
3. If a facility admin approves the carrier
3.1 Set `status` for `carrierfac` to `ok`
3.2 `carrier.fac_count`++
3.3 if `carrier.fac_count` > 0, set `status` for `carrier` to `ok` (iow: make the object visible)
4. If a facility admin or carrier operator revokes/removes a `carrierfac` object
4.1. `carrier.fac_count`--
4.2 If `carrier.fac_count` == 0, set `status` for `carrier` to `pending` (iow: make the object invisible)

**Comment by mustangthz** [comment]
+1On Mar 7, 2023, at 7:18 PM, Arnold Nipper ***@***.***> wrote:﻿

I don't think it should auto approve. I think it should send an email to the FAC admins to approve.

I should have been more verbose. There are two steps.

create the carrier object
make the carrierobject visible/invisible

My proposal is to fully automate the carrier object handling

If a carrier request comes in, auto-create the object, but keep it invisible as its fac_count is 0. Iow: status is pending
If a carrier operator adds a presence to a facility
2.1 create a carrierfac object but keep it invisible, Iow: statusis pending
2.2 send a notification to all facility admins
If a facility admin approves the carrier
3.1 Set status for carrierfac to ok
3.2 carrier.fac_count++
3.3 if carrier.fac_count > 0, set status for carrier to ok (iow: make the object visible)
If a facility admin or carrier operator revokes/removes a carrierfac object
4.1. carrier.fac_count--
4.2 If carrier.fac_count == 0, set status for carrier to pending (iow: make the object invisible)

**Comment by peterhelmenstine** [comment]
PeeringDB auto approves networks into facilities today. Why would a carrier be any different? Not all facility operators (including myself) know who's in our faculties and I don't know why we would restrict that. If a carrier network wants to state they are in a building then let them.

**Comment by Juimar** [comment]
In my opinion both network in facilities and carrier in facilities should be approved by Facility owner.

**Comment by peterhelmenstine** [comment]
What's the argument for having facilities approve networks (and now carriers) because we haven't for years and it's been fine?

**Comment by Juimar** [comment]
I guess that some Facilities don’t check these entries, and there are networks por example that don’t have presence in some Facilities……it’s only a personal opinion…

**Comment by mustangthz** [comment]
I'll change my opinion.  Carrier objects should be auto-approved.  Peter and Grizz convinced me.

**Comment by martinhannigan** [comment]
+1

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by peterhelmenstine** [comment]
To be clear:

1. Carrier object creation auto approval vs.
2. Carrier-facility object (facility approval for carriers, networks and IXes) approval into a facility.

Those are quite different.

Today:

1. Networks are not auto approved. They need to be verified
2. Addition of a network, IX or carrier is auto approved by a facility

My opinion:
1. Needs to be verified by the AC
2. <indifferent>

**Comment by martinhannigan** [comment]
Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by arnoldnipper** [comment]
FACK. 

- `carrier` needs approval. To automate, we can create a policy for carriers along the procedure for facilities.
- `carrierfac`doesn't need approval

**Comment by arnoldnipper** [comment]
> PeeringDB auto approves networks into facilities today. Why would a carrier be any different?

Because we want to use carriers to indicate that they are offering services *between* `fac`|ilities.  That's why they are different from a PDP pov. Otoh, I don't care whether it's auto-approved or not. However, the [guidelines ](https://docs.peeringdb.com/committee/admin/approval-guidelines) for @peeringdb/ac should be amended
